### PR TITLE
POC: Balloc tunable initial changes.

### PR DESCRIPTION
**Pull request by nihar1313**
Balloc tunable initial changes.

### CODE REVIEW

File: balloc/balloc.h [hunk 1]
```diff
@@ -188,6 +188,7 @@ struct m0_balloc_super_block {
 	uint32_t	bsb_bsbits;           /**< block size bits: power of 2*/
 	uint32_t	bsb_gsbits;           /**< group size bits: power of 2*/
         m0_bcount_t	bsb_groupcount;       /**< # of group */
+        m0_bcount_t	bsb_indexcount;       /**< # of group */
         m0_bcount_t	bsb_prealloc_count;   /**< nr of pre-alloc blocks */
```
Comment (huanghua78) [inline balloc/balloc.h:192]: Please align these lines.

Please update the comment of "bsb_indexcount"

File: balloc/balloc.h [hunk 2]
```diff
@@ -213,6 +214,24 @@ enum {
...
 };
 
+/*
+ * BALLOC_DEF_BLOCKS_PER_GROUP * (1 << BALLOC_DEF_BLOCK_SHIFT) = 128 MB -->
+ * which equals group size in ext4
+ */
+enum {
+	/** @todo should equal to size of HDD */
+	BALLOC_DEF_CONTAINER_SIZE	= 4096ULL * 32 * 1024 * 1000,
+	BALLOC_DEF_BLOCK_SHIFT		= 12,// 4K Blocks
```
Comment (huanghua78) [inline balloc/balloc.h:224]: No C++ style comments.

File: balloc/balloc.h [hunk 3]
```diff
@@ -355,8 +350,8 @@ M0_INTERNAL void m0_balloc_init(struct m0_balloc *cb);
 M0_INTERNAL int m0_balloc_create(uint64_t              cid,
 				 struct m0_be_seg     *seg,
 				 struct m0_sm_group   *grp,
-				 struct m0_balloc    **out,
-				 const struct m0_fid  *fid);
+				 struct m0_ad_balloc_format_req *cfg,
+				 struct m0_balloc    **out);
```
Comment (huanghua78) [inline balloc/balloc.h:354]: Please align the parameter's names.

Comment (nihar1313) [reply balloc/balloc.h]: Done

File: balloc/balloc.h (hunk 2 above)
Comment (nihar1313) [reply balloc/balloc.h]: Done

File: balloc/balloc.h [hunk 4]
```diff
@@ -187,8 +187,9 @@ struct m0_balloc_super_block {
         m0_bcount_t	bsb_groupsize;        /**< group size in blocks */
```
Comment (huanghua78) [inline balloc/balloc.h:187]: It is not related to your patch, but please fix the indentation issue for this struct.
Some of the lines are indented with "TAB" (that is correct). Some of them are using 8 spaces.

File: ioservice/storage_dev.c [hunk 5]
```diff
@@ -292,7 +295,8 @@ static int storage_dev_update_by_conf(struct m0_storage_dev  *dev,
 
 	M0_PRE(m0_ref_read(&dev->isd_ref) == 1);
 	m0_storage_dev_detach(dev);
-	rc = m0_storage_dev_new_by_conf(storage_devs, sdev, false, &dev_new);
+	rc = m0_storage_dev_new_by_conf(storage_devs, sdev,
+				        group_nr, index_nr, false, &dev_new);
```
Comment (huanghua78) [inline ioservice/storage_dev.c:299]: Can you explain why this change is needed?

File: balloc/balloc.c [hunk 6]
```diff
@@ -156,6 +156,13 @@ static bool is_spare(uint64_t alloc_flags);
 static bool is_normal(uint64_t alloc_flags);
 static bool is_any(uint64_t alloc_flag);
 
+static int index_hash(struct m0_balloc_super_block *sb,
+			m0_bcount_t block_offset)
+{
+	int grp_no;
+        grp_no = (block_offset - 1) / sb->bsb_groupsize;
```
Comment (huanghua78) [inline balloc/balloc.c:163]: Indentation issue.

More, maybe we don't need to declare variable "grp_no" at all.

File: balloc/balloc.c [hunk 7]
```diff
@@ -547,9 +560,9 @@ static int sb_update(struct m0_balloc *bal, struct m0_sm_group *grp)
 	return M0_RC(rc);
 }
 
-static int balloc_sb_write(struct m0_balloc            *bal,
-			   struct m0_balloc_format_req *req,
-			   struct m0_sm_group          *grp)
+static int balloc_sb_write(struct m0_balloc		  *bal,
```
Comment (huanghua78) [inline balloc/balloc.c:563]: Usually, we only use "TAB" at the beginning of a line to do indentation.
But in line itself, we use spaces to align the indentifiers.

File: balloc/balloc.c [hunk 8]
```diff
@@ -2978,30 +3008,38 @@ static const struct m0_ad_balloc_ops balloc_ops = {
...
+	for (i = 0; i < indexcount; i++) {
+		rc = M0_BE_OP_SYNC_RET(op,
+		       m0_be_btree_create(&bal->cb_db_group_extents[i], tx, &op,
 				  &M0_FID_TINIT('b',
 						M0_BBT_BALLOC_GROUP_EXTENTS,
 						fid->f_key)),
-		       bo_u.u_btree.t_rc);
-	if (rc != 0)
-		return M0_ERR(rc);
+						bo_u.u_btree.t_rc);
```
Comment (huanghua78) [inline balloc/balloc.c:3026]: The indentation here is not right.
```
3020         for (i = 0; i < indexcount; i++) {
3021                 rc = M0_BE_OP_SYNC_RET(op,
3022                        m0_be_btree_create(&bal->cb_db_group_extents[i], tx, &op,
3023                                           &M0_FID_TINIT('b',
3024                                           M0_BBT_BALLOC_GROUP_EXTENTS,
3025                                           fid->f_key)),
3026                                        bo_u.u_btree.t_rc);
3027                 if (rc != 0)
3028                         goto out;
3029 
3030         }
```
The arguments should be aligned to the function call.
Arguments of a function, should be aligned together.

File: stob/ad.c [hunk 9]
```diff
@@ -378,6 +428,8 @@ static int stob_ad_domain_init(struct m0_stob_type *type,
 	bool                       balloc_inited;
 	int                        rc = 0;
```
Comment (huanghua78) [inline stob/ad.c:430]: Remove the blank line here, and align the identifier "bcfg" with others.

File: stob/ad.c [hunk 10]
```diff
@@ -497,6 +544,8 @@ static int stob_ad_domain_create(struct m0_stob_type *type,
 	struct m0_buf             seg0_data;
 	int                       rc;
 
+	struct m0_ad_balloc_format_req bcfg;
```
Comment (huanghua78) [inline stob/ad.c:547]: The same as above.

File: ioservice/storage_dev.c (hunk 5 above)
Comment (nihar1313) [reply ioservice/storage_dev.c:299]: m0_storage_dev_new_by_conf() call storage_dev_new().  storage_dev_new() now required device size, and extra parameter as number of groups per device and number of index(btree). Ideally this should be attribute in sdev. Now i am sending as parameter. in future, we can do above change.

File: balloc/balloc.h (hunk 4 above)
Comment (nihar1313) [reply balloc/balloc.h]: Done..

File: balloc/balloc.c (hunk 6 above)
Comment (nihar1313) [reply balloc/balloc.c]: Done

File: balloc/balloc.c (hunk 7 above)
Comment (nihar1313) [reply balloc/balloc.c]: Done

File: balloc/balloc.c (hunk 8 above)
Comment (nihar1313) [reply balloc/balloc.c]: Changes has been done. please have al ook.

File: stob/ad.c (hunk 9 above)
Comment (nihar1313) [reply stob/ad.c]: Done

File: stob/ad.c (hunk 10 above)
Comment (nihar1313) [reply stob/ad.c]: Done

File: balloc/balloc.c [hunk 11]
```diff
@@ -640,9 +652,13 @@ balloc_group_write_credit(struct m0_balloc               *bal,
                           struct balloc_groups_write_cfg *bgs,
                           struct m0_be_tx_credit         *credit)
 {
-	m0_be_btree_insert_credit(&bal->cb_db_group_extents, 2,
-		M0_MEMBER_SIZE(struct m0_ext, e_start),
-		M0_MEMBER_SIZE(struct m0_ext, e_end), credit);
+	int i;
+
+	for (i = 0; i < bal->cb_sb.bsb_indexcount; i++) {
```
Comment (huanghua78) [inline balloc/balloc.c:657]: I have a doubt about this credit calculation.
I am not sure if all the btree (cb_db_group_extents[]) are needed.
Please review the code.

File: balloc/balloc.c [hunk 12]
```diff
@@ -3012,26 +3046,29 @@ M0_INTERNAL void m0_balloc_init(struct m0_balloc *cb)
...
-M0_INTERNAL int m0_balloc_create(uint64_t              cid,
-				 struct m0_be_seg     *seg,
-				 struct m0_sm_group   *grp,
-				 struct m0_balloc    **out,
-				 const struct m0_fid  *fid)
+M0_INTERNAL int m0_balloc_create(uint64_t			 cid,
+				struct m0_be_seg		*seg,
+				struct m0_sm_group		*grp,
+				struct m0_ad_balloc_format_req	*bcfg,
+				struct m0_balloc		**out)
```
Comment (huanghua78) [inline balloc/balloc.c:3053]: Please fix the alignment.

File: balloc/balloc.c (hunk 11 above)
Comment (nihar1313) [reply balloc/balloc.c:657]: Please correct me if i am wrong here, Following function balloc_group_write_credit() will be called during mkfs, this is onetime job, where we need to create multiple btrees[No of indexes]. We need  indexcount * No of btree credits here. where btree db formatting will happen.

File: balloc/balloc.c (hunk 12 above)
Comment (nihar1313) [reply balloc/balloc.c:3053]: Taken care..

**Review by vidyadhar-pinglikar — APPROVED** [review]

**Comment by nihar1313** [comment]
The new changes has been reviewed and merged into balloc_intigration branch. closing this.

**Pull request closed without merging**